### PR TITLE
Add cards/tiles to the users/monthly pages.

### PR DESCRIPTION
**Pull request by peregrineshahin**
Consistent with Home page cards UX.

### CODE REVIEW

**Comment by peregrineshahin** [comment]
Follow up of <https://github.com/glinscott/fishtest/pull/1711#issuecomment-1565979074>

**Comment by ppigazzini** [comment]
DEV updated, I would like @dav1312 review :)

**Comment by vdbergh** [comment]
All cards are 1 line except this one:

Screenshot from 2023-05-29 11-28-42

I also think that perhaps a 2x3 table will look better.

**Comment by peregrineshahin** [comment]
smaller font will do it

**Comment by ppigazzini** [comment]
What about moving "years" in the title?
"CPU time (years)"

**Comment by peregrineshahin** [comment]
> What about moving "years" in the title? "CPU time (years)"

Brilliant.

**Comment by ppigazzini** [comment]
Brilliant if the number is always computed in years, though.

**Comment by peregrineshahin** [comment]
> Brilliant if the number is always computed in years, though.

it is always in years
years is already hard coded

**Comment by peregrineshahin** [comment]
> I also think that perhaps a 2x3 table will look better.

would have a bit of an empty space or bloated boxs I tried them locally, doesn't look good

**Comment by ppigazzini** [comment]
> > I also think that perhaps a 2x3 table will look better.
> 
> would have a bit of an empty space or bloated boxs I tried them locally, doesn't look good

On mobile/portrait the distribution of the cards is in 3 rows 2 columns, this could be another issue for a 2x3 table.

**Comment by ppigazzini** [comment]
DEV updated, it looks good on my mobile/portrait.

File: server/fishtest/templates/users.mak [hunk 1]
```diff
@@ -15,48 +15,66 @@
   % endif
 </h2>
 
-<div class="row" style="padding: 1em 0">
-  <div class="col-sm">
-    <div class="row">
-      <div class="col text-end"><b>Testers</b></div>
-      <div class="col text-start">
-        ${sum(u['str_last_updated'] != 'Never' for u in users)}
+<div class="row row g-3 mb-3">
```
Comment (dav1312) [inline server/fishtest/templates/users.mak:18]: Duplicated row class?

Comment (peregrineshahin) [reply server/fishtest/templates/users.mak]: that's unintended, fixed.

File: server/fishtest/templates/users.mak [hunk 2]
```diff
@@ -15,48 +15,66 @@
...
-    <div class="row">
-      <div class="col text-end"><b>CPU time</b></div>
-      <div class="col text-start">
-        ${f"{sum(u['cpu_hours'] for u in users)/(24*365):.2f} years"}
+  </div>
+  <div class="col-6 col-sm">
+    <div class="card card-lg-sm text-center">
+      <div class="card-header text-nowrap" title="CPU time">CPU time (years)</div>
+      <div class="card-body">
+        <h4 class="card-title mb-0 monospace" style="font-size: 20px;">
```
Comment (dav1312) [inline server/fishtest/templates/users.mak:75]: With the "years" removed u can also try removing the hardcoded font-size

**Comment by ppigazzini** [comment]
DEV updated.

File: server/fishtest/templates/users.mak (hunk 2 above)
Comment (peregrineshahin) [reply server/fishtest/templates/users.mak:75]: well, that's intended but as you wish.

**Comment by ppigazzini** [comment]
DEV update after https://github.com/glinscott/fishtest/commit/3b0689090ba7643dde7743b5eb7f1152d5a30752

**Comment by vdbergh** [comment]
Perhaps "CPU years" (instead of "CPU time (years)")?

**Comment by ppigazzini** [comment]
I propose to sort the tiles like the table headers, so
Testers"; "Developers"; "Active testers"; "CPU time (years)"; "Games played", "Tests submitted";

**Comment by ppigazzini** [comment]
DEV updated after https://github.com/glinscott/fishtest/pull/1713/commits/47318f4311aea73c12affecfd303e885b0d40805

**Comment by peregrineshahin** [comment]
> DEV updated after [47318f4](https://github.com/glinscott/fishtest/commit/47318f4311aea73c12affecfd303e885b0d40805)

Looks good to me.

**Comment by peregrineshahin** [comment]
On a different note I'm looking at the tables and some users have last updated Never which we use the negation of that to determine the number of testers. could you please explain the logic?

**Comment by peregrineshahin** [comment]
How sg last active: never?
Capture

**Comment by peregrineshahin** [comment]
Okay now I understand, it refers to the last_active for a worker not user.
But still linrock is not considered tester bc he does both?

**Comment by ppigazzini** [comment]
It's the latest `task["time_updated"]` for all runs, so developers that never contributed CPU have "Never".

https://github.com/glinscott/fishtest/blob/d5f26eb5257b8e5d2720515b1028d7c7a8b8e26b/server/utils/delta_update_users.py#L227-L236

https://github.com/glinscott/fishtest/blob/d5f26eb5257b8e5d2720515b1028d7c7a8b8e26b/server/utils/delta_update_users.py#L146-L157

**Comment by ppigazzini** [comment]
> Okay now I understand, it refers to the last_active for a worker not user. But still linrock is not considered tester bc he does both?

linrock should be counted as "Tester" and "Developer"

https://github.com/glinscott/fishtest/blob/d5f26eb5257b8e5d2720515b1028d7c7a8b8e26b/server/fishtest/templates/users.mak#L21-L23

https://github.com/glinscott/fishtest/blob/d5f26eb5257b8e5d2720515b1028d7c7a8b8e26b/server/fishtest/templates/users.mak#L27-L30

**Comment by peregrineshahin** [comment]
> linrock should be counted in both "Tester" and "Developer" classes.
Linrock does not have `Never` so how is he counted in testers?

**Comment by peregrineshahin** [comment]
What I mean is some people run their workers from their user account, and if they do that they are not counted as testers

**Comment by peregrineshahin** [comment]
Okay, my bad it's the opposite a tester is someone that has last_updated?
What does that mean?

A tester is a person who contributes CPU time?
Why a tester is not the opposite?

**Comment by peregrineshahin** [comment]
it makes no sense. Developers are the testers. not the cpu contributers

**Comment by ppigazzini** [comment]
A developer writes a patch on his pc, opens a PR on GithHub SF, submits a run on fishtest testing framework, fishtest distributes the tasks of the run to the workers, the worker starts cutechess to play a match and get a result, the worker send the match results to fishtest, fishtest update the SPRT/NumGames result and check the exit conditions (tests).
So the only tester is fishtest.

**Comment by peregrineshahin** [comment]
Okay between you and me: Changing Testers to CPU contributors?

**Comment by peregrineshahin** [comment]
Okay never mind, now everybody knows what testers refer to in fishtest thanks to the debate :P
feel free to merge as is.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
PROD updated, thank @peregrineshahin :)

**Pull request merged**